### PR TITLE
v5.12

### LINKED ISSUE #101: qt5.12
I wonder what people think about updating this feedstock to qt5.12. 
We see different problems with qt5.6.2 on osx and linux (freecad-feedstock) and think updating qt would solve some of them. So we are definitely interested into qt5.12.

I know qt is one of the most difficult packages in conda-forge, but it's more likely to get done if more people are interested in this task. So maybe other maintainers interested in qt5.12 can show up here and give their opinon about how and when this conda-forge qt5.12 should happen. I guess  one of the main interests of anaconda to provide qt is spyder. So maybe asking @ccordoba12 what he thinks about updating this feedstock.

@mingwandroid has already build a package of qt5.12 for osx. So maybe he can give some insights on what has to be done for the other platforms.

### PR DESCRIPTION
**Pull request by isuruf**
Checklist
* [ ] Used a fork of the feedstock to propose changes
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

Fixes #75
Fixes #101

### CODE REVIEW

**Comment by ccordoba12** [comment]
Thanks a lot for this @isuruf!!

What's the plan here? Are you working on finishing the Linux build?

**Comment by isuruf** [comment]
I've got a built package at https://anaconda.org/isuruf/qt/files if you want to try it out

File: recipe/bld.bat [hunk 1]
```diff
@@ -107,7 +107,9 @@ echo QT_PATCH_VERSION = %QTVERPAT% >> Makefile
 echo CXX = cl>>Makefile
 :: This is a hack, CFLAGS_CRT becomes part of CFLAGS_EXTRA, overwriting it.
 echo CFLAGS_CRT = -DQT_LIBINFIX=\"%QT_LIBINFIX%\" >> Makefile
+echo EXTRA_CXXFLAGS =>>Makefile
```
Comment (ccordoba12) [inline recipe/bld.bat]: jom is not detecting this Makefile. That's why the Windows build is failing.

Comment (ccordoba12) [reply recipe/bld.bat]: Sorry, it's the release target that jom doesn't find.

**Comment by ocefpaf** [comment]
This is passing on macOS and times out on Linux! Awesome work @isuruf! I'm testing the Linux packages locally today and I'll report back. (Mostly I'll install this with `matplotlib` and `qgis` and check if they work properly.)

**Comment by mingwandroid** [comment]
I never got as far as completing the linux build here, ISTR getting close, "close but no biscuit" as they say in the UK sometimes.

Anyway, thank you so much for doing this @isuruf! I'm not making promises I'll not be able to keep, but I do have a 2nd hand powerful Xeon server here and I will see if I can look into Windows a bit.

@isuruf, thanks again!

**Comment by mingwandroid** [comment]
Oops. times out? OK, I'll see about docker on this server and see if I can kick at Linux some

**Comment by ocefpaf** [comment]
> Oops. times out? OK, I'll see about docker on this server and see if I can kick at Linux some

Don't worry about Linux. @isuruf and I can build that locally. If you have some spare time take a look at the Windows builds b/c that is harder for us.

**Comment by isuruf** [comment]
Currently, qdoc needs the `clangdev` package in runtime, but I'm splitting clangdev into several packages including `libclang` at https://github.com/conda-forge/clangdev-feedstock/pull/54

File: recipe/build.sh [hunk 2]
```diff
@@ -127,16 +114,55 @@ if [[ ${HOST} =~ .*linux.* ]]; then
...
+      COS6_MISSING_DEFINES["BTN_TRIGGER_HAPPY17"]="0x2d0"
+      COS6_MISSING_DEFINES["INPUT_PROP_POINTER"]="0x00"
+      COS6_MISSING_DEFINES["INPUT_PROP_DIRECT"]="0x01"
+      COS6_MISSING_DEFINES["INPUT_PROP_BUTTONPAD"]="0x02"
+      COS6_MISSING_DEFINES["INPUT_PROP_SEMI_MT"]="0x03"
+      COS6_MISSING_DEFINES["INPUT_PROP_MAX"]="0x1f"
+      COS6_MISSING_DEFINES["INPUT_PROP_CNT"]="0x20"
+      COS6_MISSING_DEFINES["ABS_MT_SLOT"]="0x2f"
+      COS6_MISSING_DEFINES["ABS_MT_PRESSURE"]="0x3a"
+      COS6_MISSING_DEFINES["ABS_MT_DISTANCE"]="0x3b"
```
Comment (isuruf) [inline recipe/build.sh]: These values need to be checked carefully against `/usr/include/linux/input-events-codes.h` from a recent linux distro.

File: recipe/patches/0013-qtwebengine-missing_EVIOCGPROP.diff [hunk 3]
```diff
@@ -0,0 +1,26 @@
+--- qtwebengine/src/3rdparty/chromium/third_party/libevdev/src/src/libevdev.c.bak	2019-07-24 15:16:31.791915512 -0500
++++ qtwebengine/src/3rdparty/chromium/third_party/libevdev/src/src/libevdev.c	2019-07-24 15:17:55.956287549 -0500
+@@ -23,6 +23,10 @@
+ #define EVIOCGMTSLOTS(len)  _IOC(_IOC_READ, 'E', 0x0a, len)
+ #endif
+ 
++#ifndef EVIOCGPROP
++#define EVIOCGPROP(len)    _IOC(_IOC_READ, 'E', 0x09, len)
```
Comment (isuruf) [inline recipe/patches/0013-qtwebengine-missing_EVIOCGPROP.diff]: Same here from `/usr/include/linux/input.h`

File: recipe/build.sh (hunk 2 above)
Comment (mingwandroid) [reply recipe/build.sh]: While I agree it's good to check and that we should, isn't it the case that the Linux kernel promises not to break userspace. That would be what changing any of these values is to my mind?

**Comment by mingwandroid** [comment]
OK, @ocepaf, I'll try to look at Windows  now.

File: recipe/build.sh (hunk 2 above)
Comment (isuruf) [reply recipe/build.sh]: I meant when reviewing this PR. I might have made typos.

**Comment by isuruf** [comment]
Waiting for https://github.com/AnacondaRecipes/aggregate/pull/164 to merge https://github.com/conda-forge/clangdev-feedstock/pull/55 and get libclang.

**Comment by ocefpaf** [comment]
> Waiting for [AnacondaRecipes/aggregate#164](https://github.com/AnacondaRecipes/aggregate/pull/164) to merge [conda-forge/clangdev-feedstock#55](https://github.com/conda-forge/clangdev-feedstock/pull/55) and get libclang.

I know that @msarahan is working on https://github.com/AnacondaRecipes/aggregate/pull/164 and that should be rebuild soon. Let's keep an eye for it.

**Comment by ccordoba12** [comment]
@isuruf, please change the build directory on Windows from `D:\` to `C:\` to get more space for the build. That's always changed when re-rendering the feedstock (like in PR #108):

https://github.com/conda-forge/qt-feedstock/pull/108/files#diff-3bdc3996454bb34a0d6efc35493f5cb8L15

**Comment by isuruf** [comment]
I've changed it but this PR would time out in azure anyway.

**Comment by ccordoba12** [comment]
Sure, it's just to see if other errors appear during the build before it times out.

**Comment by isuruf** [comment]
Thanks to @msarahan for providing me with a 8 core windows machine to debug and for the sagemath project for the 48 core linux machine. I've built the linux and windows packages and uploaded them.

(Windows packages were built with VS 2019 with VS 2017 toolset activated. Had to hack the scripts from vs2017_win-64 manually, but it should work.)

**Comment by ocefpaf** [comment]
@isuruf this is awesome! I'll put a migrator for the new `qt` later today.

**Comment by dhirschfeld** [comment]
Wow - this is awesome! Thanks all for the huge effort getting this over the line! 🚀

**Pull request merged**